### PR TITLE
tunnel: T3030: Add erspan protocol support

### PR DESCRIPTION
**Pull request by jack9603301**

### CODE REVIEW

File: interface-definitions/interfaces-tunnel.xml.in [hunk 1]
```diff
@@ -283,6 +291,70 @@
                   </leafNode>
                 </children>
               </node>
+              <node name="erspan">
```
Comment (c-po) [inline interface-definitions/interfaces-tunnel.xml.in]: You open up a new node `erspan` for every possible tunnel, this makes not much sense. Please move this under `set interfaces tunnel tun0 parameters`.

File: python/vyos/ifconfig/__init__.py [hunk 2]
```diff
@@ -39,6 +39,8 @@
 from vyos.ifconfig.tunnel import IP6IP6If
 from vyos.ifconfig.tunnel import SitIf
 from vyos.ifconfig.tunnel import Sit6RDIf
+from vyos.ifconfig.tunnel import ERSpanIf
+from vyos.ifconfig.tunnel import ip6ERSpanIf
```
Comment (c-po) [inline python/vyos/ifconfig/__init__.py]: Plese use ER6SpanIf instead as all the other tunnels do.

File: python/vyos/ifconfig/tunnel.py [hunk 3]
```diff
@@ -244,3 +247,218 @@ def _create(self):
...
+
+    default = {'type': 'erspan'}
+    options = ['local', 'remote', 'dev', 'key', 'erspan_ver']
+    
+    def _create(self):
+        # do not call _Tunnel.create, building fully here
+        
+        conf = Config()
+        base = ['interfaces', 'tunnel']
+        tunnel = get_interface_dict(conf, base)
```
Comment (c-po) [inline python/vyos/ifconfig/tunnel.py]: We do not query the configuration backend from a backend script. This must be done in the interfaces-tunnel.py frontend. Layers should not be intermixed.

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
Thank you for picking this up so fast. I see two main obstacles in this PR.

- You twist the implementation and start calling the config backend from a library script which just wraps iproute2 and kernel calls - we never required this and I see no reason why we should now. You must read in configuration from `interfaces-tunnel.py` and then pass the dict down to `_Tunnel()` or one of its derived classes.

- You configure the input interface which is to be mirrored unter the ERSPAN interface, this does not feel right. We have `set interfaces ethernet eth0 mirror <interface>` for this, so the erspan tunnel only acts as the encapsulator for the mirrored traffic. The source which is used for mirroring should still be configured by `set interfaces ethernet eth0 mirror <interface>` where <interface> can be e.g. `tun0`

Idea: should we probably create a new erspan interface level instead @dmbaturin something like `set interfaces erspan erspan0`

Please also provide a smoketest for the implementation and a configuration how this can be tested in real life by also setting up a regular Linux host as ERSPAN receiver which is used for traffic analysis.

**Comment by jack9603301** [comment]
It seems that I need to create an independent command implementation for erspan, which may be more convenient. In addition, I need to improve the support of port mirroring (the existing port mirroring only supports a single direction and needs to consider providing bidirectional support)

**Comment by jack9603301** [comment]
Is there a way to port the following commands to vyos-1x (so I can provide full port mirroring support)

```
set interfaces ethernet eth0 mirror
```

**Comment by c-po** [comment]
The `mirror` action could be extracted in a different PR from https://github.com/vyos/vyatta-cfg-qos/blob/current/scripts/vyatta-qos.pl#L322-L346 & https://github.com/vyos/vyatta-cfg-qos/blob/current/templates/interfaces/ethernet/node.tag/mirror/node.def and implemented in https://github.com/vyos/vyos-1x/blob/current/python/vyos/ifconfig/ethernet.py as it feels `mirror` is only available for ethernet but there is `redirect` for other interfaces types. I do not know the difference unfortunately.

**Comment by jack9603301** [comment]
I would like to ask a question, whether it is technically possible to directly port its interface definition to the interface definition of vyos-1x

**Comment by c-po** [comment]
Yes it is - it would consist of two PRs - one adding it in vyos-1x and one removing it from vyatta-cfg-qos.

**Comment by jack9603301** [comment]
Thank you. I'll do this gradually. Thank you for your answer. (I might consider implementing full port mapping as a separate subtask from erspan.)

File: src/conf_mode/interfaces-erspan.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    """
+    if config:
+        conf = config
+    else:
+        conf = Config()
+    base = ['interfaces', 'erspan']
+    erspan = get_interface_dict(conf, base)
+    
+    if 'mtu' not in erspan:
+        erspan['mtu'] = '1500'
```
Comment (c-po) [inline src/conf_mode/interfaces-erspan.py]: default MTU is always 1500, this is not needed

File: src/conf_mode/interfaces-erspan.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    
+    tmp = leaf_node_changed(conf, ['encapsulation'])
```
Comment (c-po) [inline src/conf_mode/interfaces-erspan.py:52]: Why encapsulation changed? There is only erspan encapsulation.

File: python/vyos/ifconfig/erspan.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    ERSpanIf: private base class for ERSPAN Over GRE and IPv4 tunnels
+    """
+    def update(self, config):
+        ifname = config['ifname']
+        local_ip = config['local_ip']
+        remote_ip = config['remote_ip']
+        key = config['parameters']['ip']['key']
+        version = config['parameters']['version']
+        if ifname not in interfaces():
+            command = f'ip link add dev {ifname} type erspan local {local_ip} remote {remote_ip} seq key {key} erspan_ver {version}'
```
Comment (c-po) [inline python/vyos/ifconfig/erspan.py]: We do not change low level interface parameters in `update()` please properly derive `_create()``

**Review by c-po — CHANGES_REQUESTED** [review]

File: src/conf_mode/interfaces-erspan.py (hunk 5 above)
Comment (jack9603301) [reply src/conf_mode/interfaces-erspan.py:52]: Contains erspan and ip6erspan

File: python/vyos/ifconfig/erspan.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+                
+        self._cmd(command)
+    
+    def update(self, config):
+        ifname = config['ifname']
+        local_ip = config['local_ip']
+        remote_ip = config['remote_ip']
+        key = config['parameters']['ip']['key']
+        version = config['parameters']['version']
+        command = f'ip link set dev {ifname} type erspan local {local_ip} remote {remote_ip} seq key {key} erspan_ver {version}'
```
Comment (c-po) [inline python/vyos/ifconfig/erspan.py:113]: As already mentioned, we do not change low level interface parameters in `update()`. See `tunnel.py:change_options()` and `tunnel.py:_create()`.

Any reason this is not derived from the `_Tunnel()` class`

**Review by c-po — CHANGES_REQUESTED** [review]

File: python/vyos/ifconfig/erspan.py (hunk 7 above)
Comment (jack9603301) [reply python/vyos/ifconfig/erspan.py:113]: Please see if my latest code structure is what you expect

File: python/vyos/ifconfig/erspan.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        },
+    }
+    
+    options = ['local_ip','remote_ip','encapsulation','parameters']
+    
+    def __init__(self,ifname,**config):
+        self.config = deepcopy(config) if config else {}
+        super().__init__(ifname, **self.config)
+    
+    def change_options(self):
```
Comment (c-po) [inline python/vyos/ifconfig/erspan.py:54]: I wonder why there can't be a common change function as there is the `type` field in `self.config` which should either hold `erspan` or `ip6erspan`, please consult the existing implementations of gre or ipip.

The code looks too complex to me at first glance.

**Review by c-po — CHANGES_REQUESTED** [review]

File: python/vyos/ifconfig/erspan.py (hunk 8 above)
Comment (jack9603301) [reply python/vyos/ifconfig/erspan.py:54]: Hey, when I detect that an interface type has changed, I need to delete the previous interface first

Comment (jack9603301) [reply python/vyos/ifconfig/erspan.py:54]: In the existing implementation of tunnel, when the package type is changed, there is a same implementation, I just copied this implementation
https://github.com/vyos/vyos-1x/blob/0b06b4807887357b2b9eed6a35928bb343c41028/src/conf_mode/interfaces-tunnel.py#L129-L135

https://github.com/vyos/vyos-1x/blob/0b06b4807887357b2b9eed6a35928bb343c41028/src/conf_mode/interfaces-tunnel.py#L65-L66

When I try to modify erspan to ip6erspan, if I use `ip link set` to modify it, the following error will be reported:

```bash
Failed to get existing tunnel info.
```

**Comment by jack9603301** [comment]
@c-po I have already responded to this question, please take a look

**Comment by c-po** [comment]
What was changed?

**Comment by jack9603301** [comment]
@c-po Since there are some reply points for your question, and you have not yet responded, I don’t know how to make the next modification. When `erspan` is changed to `ip6erspan` encapsulation mode, the `erspan` interface must be deleted first

**Comment by jack9603301** [comment]
Modify the command line to streamline configuration (support package type automatic detection), the test passed

**Comment by jack9603301** [comment]
@c-po Request review merger

**Comment by jack9603301** [comment]
@c-po Can we consider reviewing it again

**Comment by c-po** [comment]
The XML implementation is a 99% 1:1 copy of the tunnel interface. If there are two places which share the same CLI definition this should be moved to a generic file which will be included by both consumers. Please change the XML definitions accordingly. This is also true for e.g. the `verify()` section of the Python handler. One hint, could the local and remote ip verification probably moved into a dedicated helper function which will be called both by erspan and/or tunnel? Or can there be the same Python handler code `interface-tunnel-erspan.py` which supports dealing with both types of interfaces but the both have a slightly different CLI?

Also the `dhcp-interface` implementation contains a bug (https://phabricator.vyos.net/T3188) the question is, is this really required here? I see no benefit. Please only duplicate code paths which are really required, we can always extend later on.

I also wonder if all values must be configurable or could be derived/calculated during runtime of the scripts.

**Comment by jack9603301** [comment]
> I also wonder if all values must be configurable or could be derived/calculated during runtime of the scripts.

ERSPAN not many parameters must be configured, but if the user needs to be configured, you can set, in fact, ERSPAN and don't have much available for configuration parameters, about the packaging parameters, here to imitate the realization of the tunnel in order to ensure consistency, in fact, after the encapsulation type has changed, we need to delete the ERPSAN tunnel and then to recreate the tunnel pursuant to the requirements of the packaging

> The XML implementation is a 99% 1:1 copy of the tunnel interface. If there are two places which share the same CLI definition this should be moved to a generic file which will be included by both consumers. Please change the XML definitions accordingly. This is also true for e.g. the verify() section of the Python handler. 

done

**Pull request merged**